### PR TITLE
Albrja/mic 5397/mypy artifact

### PR DESCRIPTION
**Pull request by albrja**
### Albrja/mic 5397/mypy artifact
- *Category*: Feature
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5397

### Changes and notes
-remove mypy errors from framework/artifact/artifact.py

### CODE REVIEW

File: src/vivarium/framework/artifact/artifact.py [hunk 1]
```diff
@@ -214,15 +214,15 @@ def __str__(self):
         return out
 
 
-def _to_tree(keys: List[str]) -> Dict[str, Dict[str, List[str]]]:
-    out = defaultdict(lambda: defaultdict(list))
+def _to_tree(keys: list[str]) -> dict[str, dict[str, list[str]]]:
+    out: defaultdict[str, dict[str, list[str]]] = defaultdict(lambda: defaultdict(list))
```
Comment (albrja) [inline src/vivarium/framework/artifact/artifact.py:218]: This was weird to me that I needed to type hint the args to a dict that can take any structure.

File: src/vivarium/framework/artifact/artifact.py [hunk 2]
```diff
@@ -11,12 +10,13 @@
...
 
 """
+from __future__ import annotations
 
 import re
 import warnings
 from collections import defaultdict
 from pathlib import Path
-from typing import Any, Dict, List, Optional, Union
+from typing import Any, Iterable, Union
```
Comment (stevebachmeier) [inline src/vivarium/framework/artifact/artifact.py:19]: i think we've been leaning towards cleaning up Unions and replacing w/ | notation throughout?

File: src/vivarium/framework/artifact/artifact.py [hunk 3]
```diff
@@ -43,28 +43,28 @@ def __init__(self, path: Union[str, Path], filter_terms: Optional[List[str]] = N
         self._path = Path(path)
         self._filter_terms = filter_terms
         self._draw_column_filter = _parse_draw_filters(filter_terms)
-        self._cache = {}
+        self._cache: dict[str, Any] = {}
```
Comment (stevebachmeier) [inline src/vivarium/framework/artifact/artifact.py:46]: Aren't the values just pd.DataFrames?

File: src/vivarium/framework/artifact/artifact.py [hunk 4]
```diff
@@ -193,16 +193,16 @@ def clear_cache(self) -> None:
         """
         self._cache = {}
 
-    def __iter__(self):
+    def __iter__(self) -> Iterable[str]:
```
Comment (stevebachmeier) [inline src/vivarium/framework/artifact/artifact.py:196]: Should return type `Iterator`

File: src/vivarium/framework/artifact/artifact.py [hunk 5]
```diff
@@ -214,15 +214,15 @@ def __str__(self):
         return out
 
 
-def _to_tree(keys: List[str]) -> Dict[str, Dict[str, List[str]]]:
-    out = defaultdict(lambda: defaultdict(list))
+def _to_tree(keys: list[str]) -> dict[str, dict[str, list[str]]]:
```
Comment (stevebachmeier) [inline src/vivarium/framework/artifact/artifact.py:217]: Add return type - I'm surprised mypy passed?

File: src/vivarium/framework/artifact/artifact.py (hunk 1 above)
Comment (stevebachmeier) [reply src/vivarium/framework/artifact/artifact.py:218]: Yeah, my brain hurts right now and it's hard for me to figure out why that might be (or even if what you've done is correct)

File: src/vivarium/framework/artifact/artifact.py [hunk 6]
```diff
@@ -214,15 +214,15 @@ def __str__(self):
...
+    out: defaultdict[str, dict[str, list[str]]] = defaultdict(lambda: defaultdict(list))
     for k in keys:
         key = k.split(".")
         if len(key) == 3:
             out[key[0]][key[1]].append(key[2])
         else:
             out[key[0]][key[1]] = []
-    return out
+    return dict(out)
```
Comment (stevebachmeier) [inline src/vivarium/framework/artifact/artifact.py:225]: why did you add this?

**Review by stevebachmeier — CHANGES_REQUESTED** [review]

File: src/vivarium/framework/artifact/artifact.py (hunk 3 above)
Comment (albrja) [reply src/vivarium/framework/artifact/artifact.py:46]: No you can put anything in the artifact

File: src/vivarium/framework/artifact/artifact.py (hunk 4 above)
Comment (albrja) [reply src/vivarium/framework/artifact/artifact.py:196]: You have to provide the type arg too.

File: src/vivarium/framework/artifact/artifact.py (hunk 5 above)
Comment (albrja) [reply src/vivarium/framework/artifact/artifact.py:217]: It is there?

File: src/vivarium/framework/artifact/artifact.py (hunk 6 above)
Comment (albrja) [reply src/vivarium/framework/artifact/artifact.py:225]: It wanted a dict returned and it originally gets defined as a default dict.

File: src/vivarium/framework/artifact/artifact.py (hunk 4 above)
Comment (stevebachmeier) [reply src/vivarium/framework/artifact/artifact.py:196]: I mean `__iter__(self) -> Iterator[str]:`

**Review by stevebachmeier — APPROVED** [review]
Looks good aside from the __iter__() return type

File: src/vivarium/framework/artifact/artifact.py [hunk 7]
```diff
@@ -11,12 +10,13 @@
...
 
 """
+from __future__ import annotations
 
 import re
 import warnings
 from collections import defaultdict
 from pathlib import Path
-from typing import Any, Dict, List, Optional, Union
+from typing import Any, Iterator
```
Comment (rmudambi) [inline src/vivarium/framework/artifact/artifact.py:19]: Iteror should be imported from as `from collections.abc import Iterator`

**Review by rmudambi — APPROVED** [review]
